control-plane/cni/main.go

```diff
@@ -135,56 +162,120 @@ func cmdAdd(args *skel.CmdArgs) error {
 
 	client, err := kubernetes.NewForConfig(restConfig)
 	if err != nil {
-		return fmt.Errorf("Error initializing Kubernetes client: %s", err)
+		return fmt.Errorf("error initializing Kubernetes client: %s", err)
 	}
 
 	pod, err := client.CoreV1().Pods(podNamespace).Get(ctx, podName, metav1.GetOptions{})
 	if err != nil {
 		return err
 	}
 
-	// TODO: Add tests for all of the logic below
-	// TODO: Remove has hasBeenInjected and instead look for consul.hashicorp.com/cni-proxy-config annotation
-	// TODO: Add wait and timeout for annotations to show up
-	if hasBeenInjected(*pod) {
-		// If everything is good, add an annotation to the pod
-		// TODO: Remove this as it is just a stub to prove that we can do kubernetes things with the plugin
-		annotations := map[string]string{
-			keyCNIStatus: "true",
-		}
-		pod.SetAnnotations(annotations)
-		_, err = client.CoreV1().Pods(podNamespace).Update(context.TODO(), pod, metav1.UpdateOptions{})
-		if err != nil {
-			return err
-		}
-	} else {
-		logger.Debug("skipping traffic redirect on un-injected pod")
+	// skip injecting pod that has consul.hashicorp.com/connect-inject: false
+	if skipConnectInject(*pod) {
+		logger.Debug("skipping traffic redirect on un-injected pod: %s", pod.Name)
+		return types.PrintResult(result, cfg.CNIVersion)
+	}
+
+	// check to see if the cni-proxy-config annotation exists and if not, wait, retry and backoff.
+	exists := waitForAnnotation(*pod, annotationCNIProxyConfig, uint64(retries))
+	if !exists {
+		logger.Error("could not retrieve annotation: %s on pod: %s", annotationCNIProxyConfig, pod.Name)
+		return types.PrintResult(result, cfg.CNIVersion)
```

Comment:
Should we return an error in this case?

Reply:
You are correct on all of these, we should throw errors for them. Updated to return correctly.